src/index.ts

```diff
@@ -2,152 +2,378 @@ const dotenv = require('dotenv')
+        console.log('Token is refreshed')
+      }
+      catch (e: any) {
+        console.error('Refresh Token Error:', e)
+        console.error(e.intuit_tid)
+      }
+    }
+    return this.client
+  }
+
   // Use to exchange code for token
   async handle(req: Request, res: Response): Promise<boolean> {
-    console.log(req.query.code)
-    if (req.query.realmId) {
-      this.oauthClient.createToken(req.url)
-        .then(function(authResponse:any) {
-          console.log('The Token is  '+ JSON.stringify(authResponse.getJson()));
-        })
-        .catch(function(e:any) {
-          console.error("The error message is :"+e.originalMessage);
-          console.error(e.intuit_tid);
-        });
-        const {state, realmId, code} = req.query
-        if (typeof realmId === "string") {
-          this.options.realmId = parseInt(realmId, 10)
-        }
+    if (req.route.path === this.webhookPath) {
+      this.handleWebhook(req, res)
+    } else if (req.query.realmId) { // oauth callback
+      try {
+        const authResponse = await this.oauthClient.createToken(req.url)
+        await this.storeTokenAndSetClient(authResponse)
+      }
+      catch (e: any) {
+        console.error('The error message is :',e)
+        console.error(e.intuit_tid)
+      }
+      /* Do we need this
+      const { state, realmId, code } = req.query
+      if (typeof realmId === "string") {
+        this.options.realmId = realmId
+      }*/
     } else {
-      res.send({'text':'Error'})
+      res.send({ 'text': 'Error' })
     }
     res.sendStatus(200)
     return true
   }
+
+  private async handleWebhook(req: Request, res: Response) {
+    var webhookPayload = JSON.stringify(req.body)
+    var signature = req.get('intuit-signature')
+
+    // if signature is empty return 401
+    if (!signature) {
+      return res.sendStatus(401)
+    }
+    // if payload is empty, don't do anything
+    if (!webhookPayload) {
+      return res.sendStatus(200)
+    }
+    // Validates the payload with the intuit-signature hash
+    var hash = crypto.createHmac('sha256', this.configOptions.webhooksVerifier).update(webhookPayload).digest('base64')
+    if (signature !== hash) {
+      return res.sendStatus(401)
+    }
+
+    for(const eventNotification of req.body.eventNotifications) {
+      const {realmId, dataChangeEvent: {entities}} = eventNotification
+      for(const entity of entities) {    
+        entity.action = `${entity.name}/${entity.operation}`
+        const eventsToExecute = Object.values(this.eventConfigurations)
+          .filter((e => e.options.type === entity.action))
+          for(const event of eventsToExecute) {
+            const ev: QBEvent = {
+            'realmId': realmId,
+            'name': entity.name,
+            'id': entity.id,
+            'operation': entity.operation,
+            'lastUpdated': entity.lastUpdated,
+            'action': entity.action
+          }
+          await this.app.handleEvent(event.id, ev)
```

Comment:
I would pass everything from entity to the event, it's just easier.
e.g. await this.app.handleEvent(event.id, {...ev, ... entity})